Hw06

amitator opened the pull request.

amitator committed:
added toString

amitator committed:
added getAll command, but not working yet

amitator committed:
fixing getAll

amitator committed:
fixed get's command for Book

amitator committed:
added AuthorDao interface

amitator committed:
added add method to insert books

amitator committed:
added Author commands

amitator committed:
added Publisher Commands and DAO

amitator committed:
added pom.xml to git

stvort commented on src/main/java/com/prus/library/domain/Author.java:
```diff
@@ -10,4 +10,26 @@ public Author(int author_id, String first_name, String last_name){
         this.last_name = last_name;
     }
+
+    public int getAuthor_id() {
```
Тут и далее. Имена не по соглашению (first_name -> firstName). id лучше сделать long и не final т.к. он может и меняться с не существующего на конкретный.

amitator replied:
fixed

stvort commented on src/main/java/com/prus/library/domain/Book.java:
```diff
@@ -46,4 +46,15 @@ public int getPublisher_id() {
     public int getAuthor_id() {
         return author_id;
     }
+
```
Т.к. мы оперируем объектами/сущностями то:
int publisher_id -> Publisher publisher
int author_id -> Author author

В ДЗ есть такая строка "Предусмотреть таблицы авторов, книг и жанров.". Таблицу и сущность жанров я в проекте не нашел )

amitator replied:
fixed

stvort commented on src/main/java/com/prus/library/shell/BookCommands.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    @ShellMethod("Add book: ID NAME ISBN YEAR TYPE PUBLISHER_ID AUTHOR_ID")
+    public void add(int id, String name, int isbn, int year, String type, int publisher_id, int author_id){
```
Заставлять пользователя вводить id не самый лучший выбор т.к. он этот id должен откуда-то взять и так чтобы не пересекаться с другим пользователем БД. Лучше id генерировать на уровне БД.

amitator replied:
fixed

stvort commented on src/main/java/com/prus/library/shell/BookCommands.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    @ShellMethod("Get books by YEAR")
+    public void getYear(int year){
```
getYear в переводе на русский будет получитьГод, что не совсем отражает поведение метода. Это же касается других команд.  Но хорошие название методов получатся если по соглашению соединить в одно название описания из `@ShellMethod`

amitator replied:
fixed

amitator committed:
added application.yml.
added test-data.sql
added AuthorDaoJdbcTest
added auto increment to all entyties
need to fix insertion without ID.

amitator committed:
fixed all ID to long.
fixed all names.
added second constructor to each entity.

amitator committed:
added PublisherDaoJdbcTest.
modified insert() in BoobkDaoJdbc.
modified insert() in PublisherDaoJdbc.

amitator committed:
added BookDaoJdbcTest

amitator committed:
fixing toString() for Book

amitator committed:
fixing Book-Author relations

amitator committed:
fixing Book-Author-Publisher relations

Commit:
fixed Book-Author-Publisher relations

Commit:
fixed command names

Commit:
added getBookByPublisher(String name).

stvort commented on src/main/java/com/prus/library/domain/Book.java:
```diff
@@ -1,25 +1,37 @@
-    private final int author_id;
+    private final long publisherId;
+    private final long authorId;
+    private Author author;
```
publisherId и authorId тут не нужны вообще т.к. они есть внутри author и publisher

amitator replied:
fixed

stvort commented on src/main/java/com/prus/library/dao/BookDaoJdbc.java:
```diff
@@ -5,67 +5,114 @@
+        List<Book> list = jdbc.query("SELECT * FROM books", new BookMapper());
+        list = getPublisherList(list);
+        return getAuthorsList(list);
+    }
```
Так лучше не делать. Запрос 1000 книг превращается в 2001 запрос. Как я писал лучше в запросе соединить таблицы книг, авторов и издателей (SELECT * FROM book LEFT JOIN authors ON ....)

stvort commented on src/test/java/com/prus/library/dao/PublisherDaoJdbcTest.java:
```diff
@@ -0,0 +1,58 @@
+
+
+@RunWith(SpringRunner.class)
+@SpringBootTest
```
Тут лучше использовать `@JdbcTest`

amitator replied:
fixed

amitator committed:
Changed @SpringBootTest to @JdbcTest

amitator committed:
fixing PR comments

amitator committed:
fixed book insertion

stvort commented on src/main/java/com/prus/library/dao/BookDaoJdbc.java:
```diff
@@ -17,66 +16,140 @@
+        long publisherId;
+        String publisherName = book.getPublisher().getName();
+        String publisherCountry = book.getPublisher().getCountry();
+        Publisher publisher = new Publisher(publisherName, publisherCountry);
```
Эти три строки можно заменить на Publisher publisher = book.getPublisher()

amitator replied:
fixed

stvort commented on src/main/java/com/prus/library/dao/BookDaoJdbc.java:
```diff
@@ -17,66 +16,140 @@
+            publisher = publisherDao.getByName(publisherName);
+            publisherId = publisher.getPublisherId();
+        } else {
+            jdbc.update("INSERT INTO publishers (`publisher_name`, `country`) " +
```
Лучше publisherDao.insert(publisher)

amitator replied:
fixed

stvort commented on src/main/java/com/prus/library/dao/BookDaoJdbc.java:
```diff
@@ -17,66 +16,140 @@
+            authorId = author.getAuthorId();
+        } else {
+            jdbc.update("INSERT INTO authors (`first_name`, `last_name`) " +
+                            "VALUES (?, ?)", authorFirstName, authorLastName);
```
Лучше authorDao.insert(author)

amitator replied:
fixed

stvort commented on src/main/java/com/prus/library/dao/BookDaoJdbc.java:
```diff
@@ -17,66 +16,140 @@
+        long authorId = 0;
+        String authorFirstName = book.getAuthor().getFirstName();
+        String authorLastName = book.getAuthor().getLastName();
+        Author author = new Author(authorFirstName, authorLastName);
```
Эти три строки можно заменить на Author author = book.getAuthor()

amitator replied:
fixed

stvort commented on src/main/java/com/prus/library/dao/BookDaoJdbc.java:
```diff
@@ -17,66 +16,140 @@
+        Author author = new Author(authorFirstName, authorLastName);
+
+        if (authorDao.existInDatabase(author)){
+            authorId = author.getAuthorId();
```
На данный момент внутренности authorDao.existInDatabase(author) никак не влияют на id author. Откуда он должен взяться? Для издателя в этом месте стоит получение его по имени. Тут такой нет

amitator replied:
fixed

stvort commented on src/main/java/com/prus/library/shell/AuthorCommands.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    @ShellMethod("Add author: FIRST_NAME LAST_NAME")
+    public void addAuthor(String first_name, String last_name){
```
Имена не по соглашению

amitator replied:
fixed

stvort commented on src/main/java/com/prus/library/shell/AuthorCommands.java:
```diff
@@ -0,0 +1,54 @@
+    public void getAllAuthors(){
+        List<Author> list = dao.getAll();
+        for(Author author: list){
+            System.out.println(author);
```
Если сделать возвращаемым типом команды `String` то все что мы вернем будет выведено в консоль.
Ну и можно использовать стримы: 
`dao.getAll().forEach(System.out::println);`
или если следовать рекомендации выше:
`return dao.getAll().stream().map(Author::toString).collect(Collectors.joining("\n"));`

amitator replied:
fixed

stvort commented on src/main/java/com/prus/library/dao/AuthorDaoJdbc.java:
```diff
@@ -0,0 +1,71 @@
+
+    public boolean existInDatabase(Author author){
+//        Author temp = jdbc.queryForObject("SELECT * FROM authors WHERE first_name=? AND last_name=?", new Object[] {author.getFirstName(), author.getLastName()}, new AuthorMapper());
+        List<Author> temp = getByLastName(author.getLastName());
```
Почему только по фамилии? Думаю Аркадий и Борис Стругацкие с вами бы не согласились

amitator replied:
fixed

amitator committed:
fixed PR comments

amitator committed:
fixed PR commented issues

stvort commented on src/main/java/com/prus/library/dao/BookDaoJdbc.java:
```diff
@@ -17,66 +16,143 @@
+        long publisherId;
+        Publisher publisher = book.getPublisher();
+        String publisherName = publisher.getName();
+        String publisherCountry = publisher.getCountry();
```
Все равно не ясно зачем publisherName и publisherCountry ? Почему не воспользоваться геттерами publisher? То же для автора.

amitator replied:
discussed in slack

stvort commented on src/main/java/com/prus/library/dao/BookDaoJdbc.java:
```diff
@@ -17,66 +16,143 @@
+                if (author.getFirstName().equalsIgnoreCase(authorFirstName)){
+                    authorId = a.getAuthorId();
+                }
+            }
```
Если можно проверить по имени и фамилии наличие в БД автора, то и его id тоже можно получить без запроса списка по authorLastName. То же касается кода несколькими строками ниже, но с учетом того, что можно сделать, чтобы authorDao.insert возвращал id вставленного автора (см. GeneratedKeyHolder)

amitator replied:
fixed

amitator committed:
fixing PR commented issues

stvort commented on src/main/java/com/prus/library/dao/AuthorDaoJdbc.java:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    public boolean existInDatabase(Author author){
+        if (getByFullName(author.getFirstName(), author.getLastName()) != null){
```
Тут если автор не будет найден возникнет исключение

amitator replied:
fixed

stvort commented on src/main/java/com/prus/library/dao/AuthorDaoJdbc.java:
```diff
@@ -0,0 +1,82 @@
+        if (!existInDatabase(author)) {
+            jdbc.update("INSERT INTO AUTHORS (FIRST_NAME, LAST_NAME) VALUES (?, ?)",
+                    author.getFirstName(),
+                    author.getLastName(), keyHolder);
```
keyHolder передан просто как еще один параметр запроса. Для работы с keyHolder нужен перегруженный метод с PreparedStatementCreator.Не забудьте при работе с ним передать в PreparedStatmet Statement.RETURN_GENERATED_KEYS

amitator replied:
fixed

stvort commented on src/main/java/com/prus/library/dao/AuthorDaoJdbc.java:
```diff
@@ -0,0 +1,82 @@
+                author.getLastName() +
+                " already exist in database.");
+        }
+        return keyHolder.getKey().intValue();
```
ID типа long, а возвращаете int

amitator replied:
fixed

amitator committed:
fixing PreparedStatementCreator for AuthorDaoJdbc

amitator committed:
playing with shell tests

amitator committed:
fixing insert()

Commit:
Fixed insert()

Commit:
fixed publisherDao insert()
